man/man1/manta-adm.1
.TH MANTA\-ADM 1 "2016" Manta "Manta Operator Commands"
.SH NAME
.PP
manta\-adm \- administer a Manta deployment
.SH SYNOPSIS
.PP
\fB\fCmanta\-adm cn [\-l LOG_FILE] [\-H] [\-o FIELD...] [\-n] [\-s] CN_FILTER\fR
.PP
\fB\fCmanta\-adm genconfig "lab" | "coal"\fR
.PP
\fB\fCmanta\-adm genconfig [\-\-directory=DIR] \-\-from\-file=FILE\fR
.PP
\fB\fCmanta\-adm show [\-l LOG_FILE] [\-a] [\-c] [\-H] [\-o FIELD...] [\-s] SERVICE\fR
.PP
\fB\fCmanta\-adm show [\-l LOG_FILE] [\-js] SERVICE\fR
.PP
\fB\fCmanta\-adm update [\-l LOG_FILE] [\-n] [\-y] [\-\-no\-reprovision] FILE [SERVICE]\fR
.PP
\fB\fCmanta\-adm zk list [\-l LOG_FILE] [\-H] [\-o FIELD...]\fR
.PP
\fB\fCmanta\-adm zk fixup [\-l LOG_FILE] [\-n] [\-y]\fR
.SH DESCRIPTION
.PP
The \fB\fCmanta\-adm\fR command is used to administer various aspects of a Manta
deployment.  This command only operates on zones within the same datacenter.
The command may need to be repeated in other datacenters in order to execute it
across an entire Manta deployment.
.TP
\fB\fCmanta\-adm cn\fR
Show information about Manta servers in this DC.
.TP
\fB\fCmanta\-adm genconfig\fR
Generate a configuration for a COAL, lab, or multi\-server deployment.
.TP
\fB\fCmanta\-adm show\fR
Show information about deployed services.
.TP
\fB\fCmanta\-adm update\fR
Update deployment to match a JSON configuration.
.TP
\fB\fCmanta\-adm zk\fR
View and modify ZooKeeper servers configuration.
.PP
With the exception of agents (which are not currently managed by this tool),
Manta components are deployed as SDC \fBzones\fP (also called \fBinstances\fP).
Each zone is part of a \fBservice\fP, which identifies its role in the system.
.PP
Services that are part of the Manta application include:
.TP
\fBauthcache\fP
Stores user identity information
.TP
\fBelectric\-moray\fP
Moray proxy that handles sharding using consistent hashing
.TP
\fBjobpuller\fP
Manages the archival of completed user compute jobs
.TP
\fBjobsupervisor\fP
Manages the execution of user compute jobs
.TP
\fBloadbalancer\fP
Handles SSL termination and loadbalancing for "webapi"
.TP
\fBmadtom\fP
Operational dashboard for component health
.TP
\fBmarlin\-dashboard\fP
Operational dashboard for job activity
.TP
\fBmarlin\fP
Zones used to execute end user compute tasks
.TP
\fBmedusa\fP
Manages end user interactive shell sessions
.TP
\fBmoray\fP
Key\-value store used to access PostgreSQL
.TP
\fBnameservice\fP
internal DNS nameservers and ZooKeeper nodes
.TP
\fBops\fP
Manages asynchronous operations like garbage collection, metering, and auditing
.TP
\fBpostgres\fP
PostgreSQL databases used for storing object and job metadata
.TP
\fBstorage\fP
Stores actual object data
.TP
\fBwebapi\fP
Handles end user API requests
.PP
These services are described in much more detail in the Manta Operator's Guide.
.PP
The SDC SAPI service stores configuration about the "manta" application, each of
the above services, and each instance of the above service.  The information
reported by this tool is derived from SDC's internal APIs, including SAPI (for
service configuration), CNAPI (for compute node information), VMAPI (for zone
information), NAPI (for network information), and IMGAPI (for image information)
services.
.PP
Many subcommands produce tabular output, with a header row, one data record per
line, and columns separated by whitespace.  With any of these commands, you can
use options:
.TP
\fB\fC\-H, \-\-omit\-header\fR
Do not print the header row.
.TP
\fB\fC\-o, \-\-columns FIELD[,FIELD...]\fR
Only print columns named \fB\fCFIELD\fR\&.  You can specify this option multiple times
or use comma\-separated field names (or both) to select multiple fields.  The
available field names vary by command and are described in the corresponding
command section above.  In general, the default set of field names for each
command is subject to change at any time.
.PP
Many commands also accept:
.TP
\fB\fC\-l, \-\-log_file LOGFILE\fR
Emit verbose log to LOGFILE.  The special string "stdout" causes output to be
emitted to the program's stdout.
.SH SUBCOMMANDS
.SS "cn" subcommand
.PP
\fB\fCmanta\-adm cn [\-l LOG_FILE] [\-H] [\-o FIELD...] [\-n] [\-s] [CN_FILTER]\fR
.PP
The \fB\fCmanta\-adm cn\fR subcommand is used to list SDC compute nodes being used in
the current Manta deployment in the current datacenter.  The default output is a
table with one row per compute node.  See above for information on the \fB\fC\-l\fR,
\fB\fC\-H\fR, and \fB\fC\-o\fR options.
.TP
\fB\fC\-n, \-\-oneachnode\fR
Instead of printing a table, emit a comma\-separated list of matching
hostnames, suitable for use with 
.BR sdc-oneachnode (1)'s 
\fB\fC\-n\fR option.  See also
.BR manta-oneach (1).
.TP
\fB\fC\-s, \-\-storage\-only\fR
Show only compute nodes with "storage" zones on them.
.PP
The optional \fB\fCCN_FILTER\fR string can be used to provide any substring of a
compute node's hostname, server uuid, administrative IP address, compute id, or
storage ids.  All matching compute nodes will be reported.
.PP
Available fields for the \fB\fC\-o/\-\-columns\fR option include "server_uuid", "host",
"dc" (the datacenter name), "admin_ip", "ram", "compute_id", "storage_ids",
and "kind" (which is either "storage" or "other").
.PP
Example: list basic info about all Manta CNs in this DC:
.PP
.RS
.nf
# manta\-adm cn
.fi
.RE
.PP
Example: list info about Manta CN with server uuid matching 7432ffc8:
.PP
.RS
.nf
# manta\-adm cn 7432ffc8
.fi
.RE
.PP
Example: list only storage nodes:
.PP
.RS
.nf
# manta\-adm cn \-s
.fi
.RE
.PP
Example: list only the hostnames (and omit the header):
.PP
.RS
.nf
# manta\-adm cn \-H \-o host
.fi
.RE
.PP
Example: list hostnames in form suitable for "sdc\-oneachnode \-n":
.PP
.RS
.nf
# manta\-adm cn \-n
.fi
.RE
.SS "genconfig" subcommand
.PP
\fB\fCmanta\-adm genconfig "lab" | "coal"\fR
.PP
\fB\fCmanta\-adm genconfig [\-\-directory=DIR] \-\-from\-file=FILE\fR
.PP
The \fB\fCmanta\-adm genconfig\fR subcommand generates a JSON configuration file
suitable for use with \fB\fCmanta\-adm update\fR\&.  The images used for each service are
the images configured in SAPI, which are generally the last images downloaded by
.BR manta-init (1), 
so this command is sometimes used as a shortcut for identifying
the latest images that have been fetched for each service.
.PP
When the first argument is \fB\fC"coal"\fR, the command produces a configuration
suitable for a small VM\-in\-a\-laptop deployment.  The configuration is always
emitted to stdout.
.PP
When the first argument is \fB\fC"lab"\fR, the command produces a configuration
suitable for a larger single\-server install.  The configuration is always
emitted to stdout.
.PP
The \fB\fC\-\-from\-file=FILE\fR form can be used to generate a configuration suitable for
a much larger, production\-style deployment.  \fB\fCFILE\fR is a JSON file in the format
specified below that describes the parameters of the deployment, including the
number of metadata shards and the set of availability zones, racks, and servers.
This form attempts to create a deployment that will survive failures of any
component, server, rack, or availability zone as long as sufficient servers,
racks, and availability zones are included in the input file.  Availability zone
and rack information can be omitted from the file, in which case the tool will
generate a configuration ignoring rack\-level and AZ\-level considerations.  This
tool uses a number of heuristics, and the output should be verified.
.PP
By default, the generated configuration is emitted to stdout.  With the
\fB\fC\-\-directory\fR option, the configuration will be written to files in the
specified directory named by availability zone.  This option must be used if the
servers in \fB\fCFILE\fR span more than one availability zone.
.PP
The input JSON file \fB\fCFILE\fR should contain a single object with properties:
.TP
\fB\fCnshards\fR (positive integer)
the number of database shards to create, which is usually one more than the
number of shards that are intended to store object metadata (in order to
accommodate jobs and low\-volume system metadata that's typically stored in
shard 1)
.TP
\fB\fCservers\fR (array of objects)
the list of servers available for deployment
.PP
Each element of \fB\fCservers\fR is an object with properties:
.TP
\fB\fCtype\fR (string: either \fB\fC"metadata"\fR or \fB\fC"storage"\fR)
identifies this server as a target for metadata services or storage services.
It's not strictly required that Manta services be partitioned in this way, but
this tool requires that because most production deployments use two classes of
hardware for these purposes.
.TP
\fB\fCuuid\fR (string)
the SDC compute node uuid for this server.  This must be unique within the
entire region.
.TP
\fB\fCmemory\fR (positive integer)
gigabytes of memory available on this server.  This is currently only used for
storage servers to determine the appropriate number of compute zones.
.TP
\fB\fCaz\fR (string)
(optional) availability zone.  If the value is omitted from any server, that
server is placed into a default availablity zone.
.TP
\fB\fCrack\fR (string)
(optional) arbitrary identifier for the rack this server is part of.  Racks
often represent fault domains, so the tool uses this information to attempt to
distribute services across racks.  If the value is omitted from any server,
that server is placed into a default rack.
.PP
See the Manta Operator's Guide for a more complete discussion of sizing and
laying out Manta services.
.SS "show" subcommand
.PP
\fB\fCmanta\-adm show [\-l LOG_FILE] [\-a] [\-c] [\-H] [\-o FIELD...] [\-s] SERVICE\fR
.PP
\fB\fCmanta\-adm show [\-l LOG_FILE] [\-js] SERVICE\fR
.PP
The \fB\fCmanta\-adm show\fR subcommand reports information about deployed Manta
component zones.  The default output is a table with one row per deployed zone.
See above for information on the \fB\fC\-l\fR, \fB\fC\-H\fR, and \fB\fC\-o\fR options.
.TP
\fB\fC\-a, \-\-all\fR
Show zones deployed in all datacenters associated with this Manta deployment.
By default, only zones deployed in the current datacenter are shown.  Many
fields for zones deployed in other datacenters will not be available.
.TP
\fB\fC\-c, \-\-bycn\fR
Instead of showing tabular output with one row per zone sorted by service,
group zones by the compute node on which each zone is deployed.  With
\fB\fC\-a/\-\-all\fR, all compute zones in other datacenters are grouped together, since
compute node information is not available for remote datacenters.
.TP
\fB\fC\-s, \-\-summary\fR
Instead of showing tabular output with one row per zone, show tabular output
with one row per group of zones having the same "service", "image", and
"shard" properties (or just "image", for zones to which "shard" does not
logically apply).  The count for each group is also reported.  With
\fB\fC\-j/\-\-json\fR, the same information is presented in JSON form.
.TP
\fB\fC\-j, \-\-json\fR
Instead of the default text\-based output, emit a JSON representation of the
summary information reported with the \fB\fC\-s/\-\-summary\fR command.  This format is 
suitable for use with \fB\fCmanta\-adm update\fR\&.  This option cannot be combined with
\fB\fC\-c/\-\-bycn\fR, \fB\fC\-a/\-\-all\fR, \fB\fC\-H/\-\-omit\-header\fR, or \fB\fC\-o/\-\-columns\fR, and it \fImust\fP
be combined with \fB\fC\-s/\-\-summary\fR\&.  (Future versions of this command may support
a different JSON\-based report when \fB\fC\-j/\-\-json\fR is used without
\fB\fC\-s/\-\-summary\fR\&.)  For details on the JSON format, see \fB\fCmanta\-adm update\fR\&.
.PP
If \fB\fCSERVICE\fR is specified, then only zones whose service name is \fB\fCSERVICE\fR will
be reported.
.PP
Available fields for the \fB\fC\-o/\-\-columns\fR option include:
.RS
.IP \(bu 2
\fB\fCdatacenter\fR: the name of the datacenter in which this zone is deployed
.IP \(bu 2
\fB\fCimage\fR: the uuid of the zone's image
.IP \(bu 2
\fB\fCversion\fR: the version of the zone's image
.IP \(bu 2
\fB\fCprimary_ip\fR: the primary IP address for this zone
.IP \(bu 2
\fB\fCservice\fR: the name of the service this zone is part of
.IP \(bu 2
\fB\fCshard\fR: the metadata shard number for this zone.  This is only meaningful
for "moray" and "postgres" zones.
.IP \(bu 2
\fB\fCstorage_id\fR: the internal storage id for this zone.  This is only present
for "storage" zones.
.IP \(bu 2
\fB\fCzonename\fR: the full unique identifier for this zone
.IP \(bu 2
\fB\fCzoneabbr\fR: the first 8 characters of "zonename"
.IP \(bu 2
\fB\fCgz_host\fR: the hostname of the CN on which this zone is deployed
.IP \(bu 2
\fB\fCgz_admin_ip\fR: the primary IP address for the CN on which this zone is
deployed
.IP \(bu 2
\fB\fCcount\fR (summary mode only): the number of zones having the same "service",
"image", and "shard" fields (where meaningful)
.RE
.PP
Note that the "count" field is only meaningful when \fB\fC\-s/\-\-summarize\fR is
specified.  The only other fields that are meaningful when \fB\fC\-s/\-\-sumarize\fR is
specified are "service", "image", "version", and "shard".
.PP
Example: list all Manta zones in the current DC
.PP
.RS
.nf
# manta\-adm show
.fi
.RE
.PP
Example: list zones in the current DC by compute node
.PP
.RS
.nf
# manta\-adm show \-c
.fi
.RE
.PP
Example: summarize Manta zones in the current DC
.PP
.RS
.nf
# manta\-adm show \-s
.fi
.RE
.PP
Example: list all Manta zones in all datacenters (no IP info available)
.PP
.RS
.nf
# manta\-adm show \-a
.fi
.RE
.PP
Example: show only postgres zones in the current datacenter
.PP
.RS
.nf
# manta\-adm show postgres
.fi
.RE
.SS "update" subcommand
.PP
\fB\fCmanta\-adm update [\-l LOG_FILE] [\-n] [\-y] [\-\-no\-reprovision] FILE [SERVICE]\fR
.PP
The \fB\fCmanta\-adm update\fR command updates a Manta deployment to match the JSON
configuration stored at path \fB\fCFILE\fR\&.  The JSON configuration describes the
precise number of instances that should be running for each version (image) of
each type of service on each server.  The update process will involve some
number of zone deployments, undeployments, and reprovisions.  For example, if

cmd/manta-adm.js
#!/usr/bin/env node
/* vim: set ft=javascript: */
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright (c) 2015, Joyent, Inc.
 */

/*
 * manta-adm.js: manage manta deployments.  Provides subcommands:
 *
 *     cn		show information about CNs
 *
 *     show		show information about all deployed services
 *
 *     update		given a JSON file describing the desired Manta
 *     			configuration, figure out how to get from the current
 *     			state to the desired state, and start taking those
 *     			actions
 *
 *     genconfig	generate a configuration suitable for a single-system
 *     			install on COAL or a lab machine or based on a
 *     			configuration file describing available resources
 *
 *     zk		view and manage configured nameserver instances
 *
 * NOTE: this file contains ONLY the CLI wrapper around the real functionality
 * contained in lib/adm.js.  Do NOT add deployment logic here.  It belongs in
 * the library that can eventually be consumed by other tools.
 */

var assertplus = require('assert-plus');
var bunyan = require('bunyan');
var cmdln = require('cmdln');
var cmdutil = require('cmdutil');
var jsprim = require('jsprim');
var path = require('path');
var util = require('util');
var vasync = require('vasync');
var VError = require('verror').VError;
var common = require('../lib/common');
var deploy = require('../lib/deploy');
var madm = require('../lib/adm');

var maArg0 = path.basename(process.argv[1]);

/*
 * node-cmdln interface for the manta-adm tool.
 */
function MantaAdm()
{
	cmdln.Cmdln.call(this, {
	    'name': maArg0,
	    'desc': 'Inspect and modify deployed Manta services'
	});
}

util.inherits(MantaAdm, cmdln.Cmdln);

MantaAdm.prototype.initAdm = function (opts, callback)
{
	var logstreams;

	if (opts.log_file == 'stdout') {
		logstreams = [ {
		    'level': 'debug',
		    'stream': process.stdout
		} ];
	} else if (opts.log_file) {
		logstreams = [ {
		    'level': 'debug',
		    'path': opts.log_file
		} ];
		console.error('logs at ' + opts.log_file);
	} else {
		logstreams = [ {
		    'level': 'fatal',
		    'stream': process.stderr
		} ];
	}

	this.madm_log = new bunyan({
	    'name': maArg0,
	    'streams': logstreams,
	    'serializers': bunyan.stdSerializers
	});

	this.madm_adm = new madm.MantaAdm(this.madm_log);
	this.madm_adm.loadSdcConfig(function (err) {
		if (err)
			fatal(err.message);
		callback();
	});
};

MantaAdm.prototype.finiAdm = function ()
{
	this.madm_adm.close();
};

MantaAdm.prototype.do_cn = function (subcmd, opts, args, callback)
{
	var self = this;
	var options = {};
	var selected;

	if (opts.columns) {
		selected = checkColumns(madm.cnColumnNames(), opts.columns);
		if (selected instanceof Error) {
			callback(selected);
			return;
		}

		options.columns = selected;
	}

	if (opts.storage_only)
		options.onlystorage = true;

	if (opts.oneachnode)
		options.oneachnode = true;

	if (opts.omit_header)
		options.omitHeader = true;

	if (args.length > 0)
		options.filter = args[0];

	if (args.length > 1) {
		callback(new Error('unexpected arguments'));
		return;
	}

	this.initAdm(opts, function () {
		var adm = self.madm_adm;
		adm.fetchDeployed(function (err) {
			if (err)
				fatal(err.message);

			adm.dumpCns(process.stdout, options);
			self.finiAdm();
		});
	});
};

MantaAdm.prototype.do_cn.help =
    'Show information about Manta servers in this DC.\n\n' +
    'Usage:\n\n' +
    '    manta-adm cn OPTIONS [FILTER]\n\n' +
    'Examples:\n\n' +
    '    # list basic info about all Manta CNs in this DC\n' +
    '    manta-adm cn\n\n' +
    '    # list info about Manta CN with server uuid matching 7432ffc8\n' +
    '    manta-adm cn 7432ffc8\n\n' +
    '    # list only storage nodes\n' +
    '    manta-adm cn -s\n\n' +
    '    # list only the hostnames (and omit the header)\n' +
    '    manta-adm cn -H -o host\n\n' +
    '    # list hostnames in form suitable for "sdc-oneachnode -n"\n' +
    '    manta-adm cn -n\n\n' +
    'FILTER may be any substring of the compute node\'s server uuid, \n' +
    'admin IP, hostname, compute id, or storage ids.\n\n' +
    '{{options}}\n' +
    'Available columns for -o:\n    ' + madm.cnColumnNames().join(', ');

MantaAdm.prototype.do_cn.options = [ {
    'names': [ 'omit-header', 'H'],
    'type': 'bool',
    'help': 'Omit the header row for columnar output'
}, {
    'names': [ 'log_file', 'l' ],
    'type': 'string',
    'help': 'Dump logs to this file (or "stdout")'
}, {
    'names': [ 'oneachnode', 'n' ],
    'type': 'bool',
    'help': 'Emit output suitable for "sdc-oneachnode -n"'
}, {
    'names': [ 'columns', 'o' ],
    'type': 'arrayOfString',
    'help': 'Select columns for output (see below)'
}, {
    'names': [ 'storage-only', 's' ],
    'type': 'bool',
    'help': 'Show only nodes used as storage nodes.'
}];

MantaAdm.prototype.do_genconfig = function (subcmd, opts, args, callback)
{
	var self = this;
	var fromfile = opts.from_file;

	if (fromfile) {
		if (args.length !== 0) {
			callback(new Error('unexpected arguments'));
			return;
		}
	} else if (args.length != 1 ||
	    (args[0] != 'lab' && args[0] != 'coal')) {
		callback(new Error(
		    'expected "lab", "coal", or --from-file option'));
		return;
	}

	this.initAdm(opts, function () {
		var adm = self.madm_adm;
		var func;
		var options = {
		    'outstream': process.stdout
		};

		if (args[0] == 'lab') {
			func = adm.dumpConfigLab;
		} else if (args[0] == 'coal') {
			func = adm.dumpConfigCoal;
		} else {
			assertplus.string(fromfile);
			func = adm.genconfigFromFile;
			options['filename'] = fromfile;
			options['errstream'] = process.stderr;
		}

		adm.fetchDeployed(function (err) {
			if (err)
				fatal(err.message);

			func.call(adm, options, function (serr, nwarnings) {
				if (serr)
					fatal(serr.message);

				if (nwarnings !== 0) {
					console.error('error: bailing out ' +
					    'because of at least one issue');
					process.exit(1);
				}
				self.finiAdm();
			});
		});
	});
};

MantaAdm.prototype.do_genconfig.help =
    'Generate a configuration for COAL or lab deployment or for \n' +
    'a larger deployment.\n' +
    '\n' +
    'Usage:\n' +
    '\n' +
    '    manta-adm genconfig lab\n' +
    ' or manta-adm genconfig coal\n' +
    ' or manta-adm genconfig --from-file=FILE\n';

MantaAdm.prototype.do_genconfig.options = [ {
    'names': [ 'from-file' ],
    'type': 'string',
    'helpArg': 'FILE',
    'help': 'Use server descriptions in FILE'
} ];

/*
 * manta-adm show: shows information about deployed services
 */
MantaAdm.prototype.do_show = function (subcmd, opts, args, callback)
{
	var self = this;
	var selected, filter;

	if ((opts.bycn || opts.all) && opts.summary) {
		callback(new Error('-c and -a cannot be used with -s'));
		return;
	}

	if (opts.json && (!opts.summary || opts.omit_header)) {
		callback(new Error('-j cannot be used without -s or with -H'));
		return;
	}

	if (opts.columns) {
		selected = checkColumns(madm.columnNames(), opts.columns);
		if (selected instanceof Error) {
			callback(selected);
			return;
		}
	}

	if (args.length > 1) {
		callback(new Error('unexpected arguments'));
		return;
	}

	if (args.length > 0)
		filter = args[0];

	this.initAdm(opts, function () {
		var adm, func;
		adm = self.madm_adm;
		if (!opts.summary) {
			if (opts.bycn)
				func = adm.dumpDeployedZonesByCn;
			else
				func = adm.dumpDeployedZonesByService;
		} else if (opts.json) {
			func = adm.dumpDeployedConfigByServiceJson;
		} else {
			func = adm.dumpDeployedConfigByService;
		}

		adm.fetchDeployed(function (err) {
			if (err)
				fatal(err.message);

			func.call(adm, process.stdout, {
			    'doall': opts.all,
			    'omitHeader': opts.omit_header,
			    'filter': filter,
			    'columns': opts.columns ? selected : null
			});
			self.finiAdm();
		});
	});
};

MantaAdm.prototype.do_show.help =
    'Show information about deployed services.\n\n' +
    'Usage:\n\n' +
    '    manta-adm show OPTIONS [SERVICE]\n\n' +
    'Examples:\n\n' +
    '    # list all Manta zones in the current DC\n' +
    '    manta-adm show\n\n' +
    '    # list zones in the current DC by compute node\n' +
    '    manta-adm show -c\n\n' +
    '    # summarize Manta zones in the current DC\n' +
    '    manta-adm show -s\n\n' +
    '    # list all Manta zones in all datacenters (no IP info available)\n' +
    '    manta-adm show -a\n\n' +
    '    # show only postgres zones in the current datacenter\n' +
    '    manta-adm show postgres\n\n' +
    '{{options}}\n' +
    'Available columns for -o:\n    ' + madm.columnNames().join(', ');

MantaAdm.prototype.do_show.options = [ {
    'names': [ 'all', 'a' ],
    'type': 'bool',
    'help': 'Show results from all datacenters, rather than just the local one'
}, {
    'names': [ 'bycn', 'c' ],
    'type': 'bool',
    'help': 'Show results by compute node, rather than by service.'
}, {
    'names': [ 'omit-header', 'H'],
    'type': 'bool',
    'help': 'Omit the header row for columnar output'
}, {
    'names': [ 'json', 'j' ],
    'type': 'bool',
    'help': 'Show results in JSON form suitable for importing with "update".'
}, {
    'names': [ 'log_file', 'l' ],
    'type': 'string',
    'help': 'dump logs to this file (or "stdout")'
}, {
    'names': [ 'columns', 'o' ],
    'type': 'arrayOfString',
    'help': 'Select columns for output (see below)'
}, {
    'names': [ 'summary', 's' ],
    'type': 'bool',
    'help': 'Show summary of deployed zones rather than each zone separately.'
} ];

/*
 * manta-adm update: deploys, undeploys, and redeploys to match a desired
 * deployment specification
 */
MantaAdm.prototype.do_update = function (subcmd, opts, args, callback)
{
	var filename, service, nchanges, adm;
	var self = this;

	if (args.length === 0) {
		callback(new Error(
		    'expected filename for desired configuration'));
		return;
	}

	if (args.length > 2) {
		callback(new Error('unexpected arguments'));
		return;
	}

	filename = args[0];
	if (args.length == 2)
		service = args[1];

	if(service && (madm.mSvcNames.indexOf(service) == -1)) {
		callback(new VError('%s is not a valid manta service', service));
		return;
	}

	vasync.pipeline({
	    'funcs': [
		function initAdm(_, stepcb) {
			self.initAdm(opts, function () {
				adm = self.madm_adm;
				stepcb();
			});
		},
		function readConfig(_, stepcb) {
			adm.readConfigFromFile(filename, stepcb);
		},
		function fetchDeployed(_, stepcb) {
			adm.fetchDeployed(stepcb);
		},
		function generatePlan(_, stepcb) {
			adm.generatePlan(stepcb, service, opts.no_reprovision);
		},
		function dumpPlan(_, stepcb) {
			adm.execPlan(process.stdout, process.stderr,
			    true, function (err, count) {
				if (err) {
					stepcb(err);
					return;
				}

				nchanges = count;
				if (count > 0 && opts.dryrun)
					console.log('To apply these changes, ' +
					    'leave off -n (--dry-run).');

				stepcb();
			    });
		},
		function uconfirm(_, stepcb) {
			if (opts.dryrun || nchanges === 0 || opts.confirm) {
				stepcb();
				return;
			}

			common.confirm(
			    'Are you sure you want to proceed? (y/N): ',
			    function (proceed) {
				process.stdout.write('\n');
				if (!proceed) {
					stepcb(new Error('aborted by user'));
				} else {
					stepcb();
				}
			    });
		},
		function execPlan(_, stepcb) {
			if (opts.dryrun || nchanges === 0) {
				stepcb();
				return;
			}

			adm.execPlan(process.stdout, process.stderr,
			    false, stepcb);
		}
	    ]
	}, function (err) {
		if (err)
			fatal(err.message);
		self.finiAdm();
		callback();
	});
};

MantaAdm.prototype.do_update.help =
    'Update deployment to match a JSON configuration.\n\n{{options}}';

MantaAdm.prototype.do_update.options = [ {
    'names': [ 'log_file', 'l' ],
    'type': 'string',
    'help': 'dump logs to this file (or "stdout")',
    'default': '/var/log/manta-adm.log'
}, {
    'names': [ 'dryrun', 'n' ],
    'type': 'bool',
    'help': 'Print what would be done without actually doing it.'
}, {
    'names': [ 'confirm', 'y' ],
    'type': 'bool',
    'help': 'Bypass all confirmations (be careful!)'
}, {
    'names': [ 'no-reprovision' ],
    'type': 'bool',
    'help': 'When upgrading a zone, always provision and deprovision ' +
	'rather than reprovision'
} ];

MantaAdm.prototype.do_zk = MantaAdmZk;

function MantaAdmZk(parent)
{
	this.mn_parent = parent;
	cmdln.Cmdln.call(this, {
	    'name': 'zk',
	    'desc': 'View and modify ZooKeeper servers configuration.'
	});
}

util.inherits(MantaAdmZk, cmdln.Cmdln);

MantaAdmZk.prototype.do_list = function (subcmd, opts, args, callback)
{
	var self = this;
	var options = {};
	var selected;

	if (args.length > 0) {
		callback(new Error('unexpected arguments'));
		return;
	}

	if (opts.columns) {
		selected = checkColumns(madm.zkColumnNames(), opts.columns);
		if (selected instanceof Error) {
			callback(selected);
			return;
		}

		options.columns = selected;
	}

	if (opts.omit_header)
		options.omitHeader = true;

	this.mn_parent.initAdm(opts, function () {
		var adm = self.mn_parent.madm_adm;
		adm.fetchDeployed(function (err) {
			var problems;

			if (err)
				fatal(err.message);
			problems = adm.dumpZkServers(process.stdout, options);
			problems.critical.forEach(function (warn) {
				console.error('error: %s', warn.message);
			});
			problems.fixable.forEach(function (warn) {
				console.error('warning: %s', warn.message);
			});

			if (problems.critical.length +
			    problems.fixable.length > 0)
				process.exit(1);
			self.mn_parent.finiAdm();
		});
	});
};

MantaAdmZk.prototype.do_list.help =
    'List configured ZooKeeper servers\n\n' +
    'Usage:\n\n' +
    '    manta-adm zk list OPTIONS\n\n' +
    'Examples:\n\n' +
    '    # list ZooKeeper servers\n' +
    '    manta-adm zk list\n\n' +
    '    # list only IPs of ZK servers\n' +
    '    manta-adm zk list --omit-header -o ip\n\n' +
    '{{options}}\n' +
    'Available columns for -o:\n    ' + madm.zkColumnNames().join(', ');

/*
 * Note that the "manta-adm" commands that may modify the system use
 * /var/log/manta-adm.log as the default log file, as those logs currently serve
 * as general debug logs.  But the "zk list" subcommand is read-only and only
 * applicable to this user, so we use a path in /var/tmp for the log.
 */
MantaAdmZk.prototype.do_list.options = [ {
    'names': [ 'omit-header', 'H'],
    'type': 'bool',
    'help': 'Omit the header row for columnar output'
}, {
    'names': [ 'log_file', 'l' ],
    'type': 'string',
    'help': 'dump logs to this file (or "stdout")',
    'default': '/var/tmp/manta-adm.log'
}, {
    'names': [ 'columns', 'o' ],
    'type': 'arrayOfString',
    'help': 'Select columns for output (see below)'
} ];

MantaAdmZk.prototype.do_fixup = function (subcmd, opts, args, callback)
{
	var self = this;
	var adm, nissues, nfixed;

	if (args.length > 0) {
		callback(new Error('unexpected arguments'));
		return;
	}

	vasync.pipeline({
	    'funcs': [
		function initAdm(_, stepcb) {
			self.mn_parent.initAdm(opts, function () {
				adm = self.mn_parent.madm_adm;
				stepcb();
			});
		},
		function fetchDeployed(_, stepcb) {
			adm.fetchDeployed(stepcb);
		},
		function dumpNameservers(_, stepcb) {
			var problems;

			console.error('CURRENT CONFIGURATION');
			problems = adm.dumpZkServers(process.stderr, {});
			if (problems.critical.length > 0) {
				problems.critical.forEach(function (e) {
					console.error('error: %s', e.message);
				});
				stepcb(new VError('bailing out after errors'));
				return;
			}

			nissues = problems.fixable.length;
			if (nissues === 0) {
				console.error('no issues to repair');
				stepcb();
				return;
			}

			console.error(
			    'The following issues should be repaired:');
			problems.fixable.forEach(function (e) {
				console.error('error: %s', e.message);
			});

			if (opts.dryrun) {
				console.error('To repair, leave off ' +
				    '-n (--dry-run)');
			}

			stepcb();
		},
		function uconfirm(_, stepcb) {
			if (opts.dryrun || nissues === 0 || opts.confirm) {
				stepcb();
				return;
			}

			common.confirm(
			    'Do you want to repair these issues now? (y/N): ',
			    function (proceed) {
				process.stdout.write('\n');
				if (!proceed) {
					stepcb(new Error('aborted by user'));
				} else {
					stepcb();
				}
			    });
		},
		function repair(_, stepcb) {
			if (opts.dryrun || nissues === 0) {
				stepcb();
				return;
			}

			adm.fixupZkServers(function (err, n) {
				if (!err)
					nfixed = n;
				stepcb(err);
			});
		}
	    ]
	}, function (err) {
		if (err)
			fatal(err.message);
		if (!opts.dryrun && nissues > 0)
			console.error('%d issue%s repaired',
			    nfixed, nfixed == 1 ? '' : 's');
		self.mn_parent.finiAdm();
		callback();
	});
};

MantaAdmZk.prototype.do_fixup.help = [
    'Repair ZooKeeper configuration',
    '',
    'This command compares the ZooKeeper configuration (defined by the ',
    'ZK_SERVERS and ZK_ID SAPI metadata properties) to the list of deployed ',
    'nameservice zones, reports any discrepancies or other issues, and ',
    'optionally repairs certain kinds of issues.  If repairs are made, only ',
    'metadata is changed.  This tool is intended for cases where a ZK server ',
    'has been undeployed and the configuration needs to be updated, or where ',
    'deployment failed and left stale configuration, or other unusual cases ',
    'where the configuration does not match the list of deployed nameservers.',
    'The "manta-adm zk list" command identifies these problem cases.',
    '',
    'Usage:',
    '',
    '    manta-adm zk [-n | --dry-run] [-y | --confirm] fixup',
    '',
    'Examples:',
    '',
    '    # check for configuration issues and repair them',
    '    manta-adm zk fixup',
    '',
    '{{options}}'
].join('\n');

MantaAdmZk.prototype.do_fixup.options = [ {
    'names': [ 'confirm', 'y' ],
    'type': 'bool',
    'help': 'Bypass all confirmations (be careful!)'
}, {
    'names': [ 'dryrun', 'n' ],
    'type': 'bool',
    'help': 'Print what would be done without actually doing it.'
}, {
    'names': [ 'log_file', 'l' ],
    'type': 'string',
    'help': 'Dump logs to this file (or "stdout")'
} ];

function checkColumns(allowed, columns)
{
	var selected, c, i;

	selected = [];
	for (i = 0; i < columns.length; i++) {
		c = columns[i].split(',');
		selected = selected.concat(columns[i].split(','));
	}

	for (i = 0; i < selected.length; i++) {
		c = selected[i];
		if (allowed.indexOf(c) == -1)
			return (new VError('unknown column: "%s"', c));
	}

	return (selected);
}

function fatal(msg)
{
	console.error('%s: %s', maArg0, msg);
	process.exit(1);
}

cmdutil.exitOnEpipe();
cmdln.main(MantaAdm);
